src/main/java/org/opensearch/knn/index/codec/KNN990Codec/KNNQuantizationStateWriter.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.KNN990Codec;
+
+import lombok.AllArgsConstructor;
+import lombok.Setter;
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.store.IndexOutput;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.quantization.models.quantizationState.QuantizationState;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Writes quantization states to off heap memory
+ */
+public final class KNNQuantizationStateWriter {
+
+    private final IndexOutput output;
+    private List<FieldQuantizationState> fieldQuantizationStates = new ArrayList<>();
+
+    /**
+     * Constructor
+     * Overall file format for writer:
+     * Header
+     * QS1 state bytes
+     * QS2 state bytes
+     * Number of quantization states
+     * QS1 field number
+     * QS1 state bytes length
+     * QS1 position of state bytes
+     * QS2 field number
+     * QS2 state bytes length
+     * QS2 position of state bytes
+     * Position of index section (where QS1 field name is located)
+     * -1 (marker)
+     * Footer
+     * @param segmentWriteState segment write state containing segment information
+     * @throws IOException exception could be thrown while creating the output
+     */
+    public KNNQuantizationStateWriter(SegmentWriteState segmentWriteState) throws IOException {
+        String quantizationStateFileName = IndexFileNames.segmentFileName(
+            segmentWriteState.segmentInfo.name,
+            segmentWriteState.segmentSuffix,
+            KNNConstants.QUANTIZATION_STATE_FILE_SUFFIX
+        );
+
+        output = segmentWriteState.directory.createOutput(quantizationStateFileName, segmentWriteState.context);
+    }
+
+    /**
+     * Writes an index header
+     * @param segmentWriteState state containing segment information
+     * @throws IOException exception could be thrown while writing header
+     */
+    public void writeHeader(SegmentWriteState segmentWriteState) throws IOException {
+        CodecUtil.writeIndexHeader(output, "KNN990Codec", 0, segmentWriteState.segmentInfo.getId(), segmentWriteState.segmentSuffix);
```

Comment:
can we use `KNNCodecVersion` to get the codecName instead of a string here, 

Also if this writer is tied to the codec, 
- should we be encapsulating that in the enum like the vector formats?
- Should this be renamed to `KNN990QuantizationStateWriter`?

Reply:
Looking at some references in Lucene I can see this is what being used:
https://github.com/apache/lucene/blob/branch_9x/lucene/core/src/java/org/apache/lucene/codecs/lucene99/Lucene99ScalarQuantizedVectorsFormat.java#L47

So we should have something like `NativeEngines99KnnVectorsFormatQSData`

Reply:
> Should this be renamed to KNN990QuantizationStateWriter

+1 on this. This should happen for both readers and writers